File: consensus/src/main/scala/co/topl/consensus/LocalChain.scala
package co.topl.consensus

import cats.data.Validated
import cats.effect.Ref
import cats.effect.kernel.Sync
import cats.implicits._
import co.topl.consensus.algebras.LocalChainAlgebra
import co.topl.models.{SlotData, TypedIdentifier}
import co.topl.typeclasses.OrderT

object LocalChain {

  object Eval {

    def make[F[_]: Sync](
      initialHead:    SlotData,
      chainSelection: OrderT[F, SlotData],
      onAdopted:      TypedIdentifier => F[Unit]
    ): F[LocalChainAlgebra[F]] =
      Ref
        .of[F, SlotData](initialHead)
        .map(headRef =>
          new LocalChainAlgebra[F] {

            def isWorseThan(newHead: SlotData): F[Boolean] =
              head.flatMap(chainSelection.compare(_, newHead).map(_ < 0))

            def adopt(newHead: Validated.Valid[SlotData]): F[Unit] =
              Sync[F].uncancelable(_ => onAdopted(newHead.a.slotId.blockId) >> headRef.update(_ => newHead.a))

            val head: F[SlotData] =
              headRef.get
          }
        )
  }
}

File: dart/consensus/lib/interpreters/local_chain.dart
import 'dart:async';

import 'package:bifrost_consensus/algebras/local_chain_algebra.dart';
import 'package:topl_protobuf/consensus/models/block_id.pb.dart';

class LocalChain extends LocalChainAlgebra {
  LocalChain(BlockId initialHead) : this._currentHead = initialHead;
  BlockId _currentHead;

  final StreamController<BlockId> _streamController =
      StreamController.broadcast();

  @override
  FutureOr<void> adopt(BlockId newHead) async {
    if (_currentHead != newHead) {
      _currentHead = newHead;
      _streamController.add(newHead);
    }
  }

  @override
  Stream<BlockId> get adoptions => _streamController.stream;

  @override
  FutureOr<BlockId> get currentHead => Future.value(_currentHead);
}
